File: Searching Algorithms/interpolation_search.py
def interpolation_search(arr, target):
    left = 0
    right = len(arr) - 1

    while(left<=right and target>=arr[left] and target<=arr[right]):
        probe = left + (right-left)*(target-arr[left])//(arr[right]-arr[left])
        if arr[probe]<target:
            left = probe+1
        elif arr[probe]>target:
            right = probe-1
        else:
            if arr[probe]==target:
                return probe
    return -1

arr = [1, 4, 7, 8, 10, 12]
print(interpolation_search(arr, 7))
print(interpolation_search(arr, 15))

File: Searching Algorithms/interpolation_search.cpp
#include <iostream>
#include <vector>
using namespace std;

int interpolation_search(vector<int> &nums, int target){
    int left = 0;
    int right = nums.size()-1;
    while (left<=right && target>=nums[left] && target<=nums[right]){
        int probe = left + (right-left)*(target-nums[left])/(nums[right]-nums[left]);
        if (nums[probe]>target){
            right = probe-1;
        }
        else if(nums[probe]<target){
            left = probe+1;
        }
        else{
            if (nums[probe]==target){
                return probe;
            }
        }
    }
    return -1;
}

int main(){
    vector<int> nums = {1, 4, 7, 8, 10, 12};
    cout<<interpolation_search(nums, 7)<<"\n";
    cout<<interpolation_search(nums, 15)<<"\n";
    return 0;
}